hack/imagebuilder/main.go

```diff
@@ -0,0 +1,296 @@
+		"latest",
+		"latest-root",
+	}
+)
+
+func init() {
+	out, err := runCmd("git", "rev-parse", "--show-toplevel")
+	if err != nil {
+		logrus.WithError(err).Error("Failed getting git root dir")
+		os.Exit(1)
+	}
+	rootDir = out
+}
+
+func init() {
+	_, err := runCmdInDir(path.Join(rootDir, "hack/tools"), "go", "build", "-o", path.Join(rootDir, "_bin/ko"), "github.com/google/ko")
+	if err != nil {
+		logrus.WithError(err).Error("Failed ensure ko")
+		os.Exit(1)
+	}
+}
+
+type options struct {
+	koDockerRepo      string
+	prowImageListFile string
+	workers           int
+	push              bool
+}
+
+func runCmdInDir(dir, cmd string, args ...string) (string, error) {
+	command := exec.Command(cmd, args...)
+	if dir != "" {
+		command.Dir = dir
+	}
+	stdOut, err := command.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+	stdErr, err := command.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+	if err := command.Start(); err != nil {
+		return "", err
+	}
+	scanner := bufio.NewScanner(stdOut)
+	var allOut string
+	for scanner.Scan() {
+		out := scanner.Text()
+		allOut = allOut + out
+		logrus.WithField("cmd", command.Args).Info(out)
+	}
+	allErr, _ := io.ReadAll(stdErr)
+	if len(allErr) > 0 {
+		logrus.WithField("cmd", command.Args).Error(string(allErr))
+	}
+	err = command.Wait()
+	return strings.TrimSpace(allOut), err
+}
+
+func runCmd(cmd string, args ...string) (string, error) {
+	return runCmdInDir(rootDir, cmd, args...)
+}
+
+type imageDef struct {
+	Dir  string `json:"dir"`
+	Arch string `json:"arch"`
+}
+
+type imageDefs struct {
+	Defs []imageDef `json:"images"`
+}
+
+func loadImageDefs(p string) ([]imageDef, error) {
+	b, err := ioutil.ReadFile(p)
+	if err != nil {
+		return nil, err
+	}
+	var res imageDefs
+	if err := yaml.Unmarshal(b, &res); err != nil {
+		return nil, err
+	}
+	return res.Defs, nil
+}
+
+func allBaseTags() ([]string, error) {
+	gitTag, err := gitTag()
+	if err != nil {
+		return nil, err
+	}
+	return append(defaultTags, gitTag), nil
+}
+
+func allTags(arch string) ([]string, error) {
+	baseTags, err := allBaseTags()
+	if err != nil {
+		return nil, err
+	}
+	if arch != allArch {
+		return baseTags, nil
+	}
+	var allTags = baseTags[:]
+	for _, arch := range otherArches {
+		for _, base := range baseTags {
+			allTags = append(allTags, fmt.Sprintf("%s-%s", base, arch))
+		}
+	}
+	return allTags, nil
+}
+
+func gitTag() (string, error) {
+	prefix, err := runCmd("date", "+v%Y%m%d")
+	if err != nil {
+		return "", err
+	}
+	postfix, err := runCmd("git", "describe", "--always", "--dirty")
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s-%s", prefix, postfix), nil
+}
+
+func setup(id *imageDef) error {
+	script := path.Join(rootDir, id.Dir, gatherStaicScriptName)
+	if _, err := os.Lstat(script); err != nil {
+		if !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+	if _, err := runCmd(script); err != nil {
+		return err
+	}
+	return nil
+}
+
+func teardown(id *imageDef) error {
```

Comment:
`setup` and `teardown` could be combined by adding a parameter.

Reply:
good idea